packages/core/src/utils/pValidators/baseValidator.ts

```diff
@@ -0,0 +1,92 @@
+import { Callable, CallableConfig } from "../../record/callable";
+
+type ValidationFunction = (input: any) => boolean;
+type ValidationResult = { isValid: boolean; errorMessage?: string };
+
+interface ValidatorRules {
```

Comment:
can you explain the motivation and use case of this `ValidatorRules`? e.g. how does this connects attributes with validators?

Reply:
validator is a class for users to validate the value of input variables, the motivation for ValidatorRules is to enable users or developers to create their own rules for validating with more flexibility

Reply:
I don't think the validator class should be complex, since it just provide a simple checking logic before running the event (for now, the event is prompt template, but there can be more events that use this validator). Maybe we can just adding static methods in this `baseValidator` and then each static method is corresponding a validator function?

The use case can be something like this: (I got reference from [Angular validator](https://github.com/angular/angular/blob/main/packages/forms/src/validators.ts#L139))

```TypeScript
class Validators {
  static min(min: number): ValidatorFn {
    return minValidator(min);
  }
}

function minValidator(min: number): ValidatorFn {
  return (control: AbstractControl): ValidationErrors|null => {
    if (isEmptyInputValue(control.value) || isEmptyInputValue(min)) {
      return null;  // don't validate empty values to allow optional controls
    }
    const value = parseFloat(control.value);
    // Controls with NaN values after parsing should be treated as not having a
    // minimum, per the HTML forms spec: https://www.w3.org/TR/html5/forms.html#attr-input-min
    return !isNaN(value) && value < min ? {'min': {'min': min, 'actual': control.value}} : null;
  };
}

const control = new FormControl(2, Validators.min(3));

console.log(control.errors); // {min: {min: 3, actual: 2}}
```

However, they used a class `FormControl` to manage their validators. This class has more complex functionalities, such as tracking the current state of the form attributes, updating validators, changing the state if the validator is not pass. 

I feel like we need to design a similar class like the `FormControl`, then instantiate this class for each input variable in the `PromptTemplate`.

For the end user, the use case of the PromptTemplate with validator looks something like this.

```TypeScript
const promptTemplate = new PromptTemplate({
  template: "I have {numOrange} oranges. Ben has 10 oranges.",
  validators: {
    numOrange: [Validators.min(10)]
  },
});
```